File: part2/Implementations/SyntaxAnalyser/Services/Tokenizer.cs
using SyntaxAnalyser.Abstractions;
using SyntaxAnalyser.Models;
using System.Text.RegularExpressions;
namespace SyntaxAnalyser.Services;

public partial class Tokenizer(string code) : ITokenizer
{
    private int _index;

    private readonly Regex[] _ignoredRules =
    [
        OnelineCommentRegex(),
        MultilinesCommentRegex(),
        WhiteSpaceRegex(),
    ];

    private readonly (Regex Regex, TokenType Type)[] _tokenRules =
    [
        (KeywordRegex(), TokenType.Keyword),
        (SymbolRegex(), TokenType.Symbol),
        (IntegerConstantRegex(), TokenType.IntegerConstant),
        (StringConstantRegex(), TokenType.StringConstant),
        (IdentifierRegex(), TokenType.Identifier),
    ];

    public Token CurrentToken
    {
        get
        {
            _index = SkipIgnored(_index);

            foreach (var (regex, type) in _tokenRules)
            {
                var current = code[_index..];
                var match = regex.Match(current);
                if (!match.Success)
                    continue;

                var token = match.Value;
                ValidateToken(type, token);
                return new Token(type, token);
            }

            var remaining = code[_index..];
            var preview = remaining.Length > 20 ? remaining[..20] : remaining;
            throw new InvalidOperationException($"Invalid token near '{preview}' at index {_index}");
        }
    }
    
    private int SkipIgnored(int index)
    {
        while (index < code.Length)
        {
            var current = code[index..];
            var matchedIgnoredToken = false;

            foreach (var ignoredRule in _ignoredRules)
            {
                var match = ignoredRule.Match(current);
                if (!match.Success)
                    continue;

                index += match.Length;
                matchedIgnoredToken = true;
                break;
            }

            if (!matchedIgnoredToken)
                break;
        }

        return index;
    }

    public bool HasMoreTokens => SkipIgnored(_index) < code.Length;
    
    public Token Advance()
    {
        var token = CurrentToken;
        _index += token.Value.Length;
        return token;
    }
    
    public Token Eat(string token)
    {
        if (HasMoreTokens &&  CurrentToken.Value == token)
        {
            return Advance();
        }
        throw new InvalidOperationException($"Invalid token near '{token}' at index {_index}");
    }
    public TokenType TokenType => CurrentToken.Type;
    public string StringValue => CurrentToken.Type == TokenType.StringConstant ?  CurrentToken.Value.Replace("\"", "") : throw new InvalidOperationException("You can't perform this action on non string value");
    public int IntValue => CurrentToken.Type == TokenType.IntegerConstant ?  int.Parse(CurrentToken.Value) : throw new InvalidOperationException("You can't perform this action on non int value");
    public string KeyWord => CurrentToken.Type == TokenType.Keyword ?  CurrentToken.Value : throw new InvalidOperationException("You can't perform this action on non keyword value");
    public char Symbol => CurrentToken.Type == TokenType.Symbol ? CurrentToken.Value.ElementAt(0) : throw new InvalidOperationException("You can't perform this action on symbol value");
    public string Identifier => CurrentToken.Type == TokenType.Identifier ? CurrentToken.Value : throw new InvalidOperationException("You can't perform this action on non identifier");

    [GeneratedRegex(@"^\d+\b")]
    private static partial Regex IntegerConstantRegex();

    [GeneratedRegex(@"^[A-Za-z_][A-Za-z0-9_]*\b")]
    private static partial Regex IdentifierRegex();

    [GeneratedRegex("""
                    ^"[^"\r\n]*"
                    """)]
    private static partial Regex StringConstantRegex();

    [GeneratedRegex(@"^(?:[{}()\[\].,;+\-*&|<>=~]|/(?![/*]))")]
    private static partial Regex SymbolRegex();

    [GeneratedRegex(@"^(class|constructor|function|method|field|static|var|int|char|boolean|void|true|false|null|this|let|do|if|else|while|return)\b")]
    private static partial Regex KeywordRegex();

    [GeneratedRegex(@"^//[^\r\n]*")]
    private static partial Regex OnelineCommentRegex();

    [GeneratedRegex(@"^/\*[\s\S]*?\*/\s*")]
    private static partial Regex MultilinesCommentRegex();

    [GeneratedRegex(@"^\s+")]
    private static partial Regex WhiteSpaceRegex();

    private void ValidateToken(TokenType type, string token)
    {
        if (type != TokenType.IntegerConstant)
            return;

        if (!int.TryParse(token, out var number) || number > 32767)
            throw new InvalidOperationException($"Integer constant out of range near '{token}' at index {_index}");
    }
}

File: part2/Implementations/submission/project10/Tokenizer.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class Tokenizer {
    private static final Pattern INTEGER_CONSTANT_PATTERN = Pattern.compile("^\\d+\\b");
    private static final Pattern IDENTIFIER_PATTERN = Pattern.compile("^[A-Za-z_][A-Za-z0-9_]*\\b");
    private static final Pattern STRING_CONSTANT_PATTERN = Pattern.compile("^\"[^\"\\r\\n]*\"");
    private static final Pattern SYMBOL_PATTERN = Pattern.compile("^(?:[{}()\\[\\].,;+\\-*&|<>=~]|/(?![/*]))");
    private static final Pattern KEYWORD_PATTERN = Pattern.compile(
        "^(class|constructor|function|method|field|static|var|int|char|boolean|void|true|false|null|this|let|do|if|else|while|return)\\b");
    private static final Pattern ONE_LINE_COMMENT_PATTERN = Pattern.compile("^//[^\\r\\n]*");
    private static final Pattern MULTI_LINE_COMMENT_PATTERN = Pattern.compile("^/\\*[\\s\\S]*?\\*/\\s*");
    private static final Pattern WHITE_SPACE_PATTERN = Pattern.compile("^\\s+");

    private static final Pattern[] IGNORED_RULES = new Pattern[] {
        ONE_LINE_COMMENT_PATTERN,
        MULTI_LINE_COMMENT_PATTERN,
        WHITE_SPACE_PATTERN
    };

    private static final Rule[] TOKEN_RULES = new Rule[] {
        new Rule(KEYWORD_PATTERN, TokenType.KEYWORD),
        new Rule(SYMBOL_PATTERN, TokenType.SYMBOL),
        new Rule(INTEGER_CONSTANT_PATTERN, TokenType.INTEGER_CONSTANT),
        new Rule(STRING_CONSTANT_PATTERN, TokenType.STRING_CONSTANT),
        new Rule(IDENTIFIER_PATTERN, TokenType.IDENTIFIER)
    };

    private final String code;
    private int index;

    Tokenizer(String code) {
        this.code = code;
    }

    Token getCurrentToken() {
        index = skipIgnored(index);

        for (Rule rule : TOKEN_RULES) {
            String current = code.substring(index);
            Matcher matcher = rule.pattern.matcher(current);
            if (!matcher.find()) {
                continue;
            }

            String token = matcher.group();
            validateToken(rule.type, token);
            return new Token(rule.type, token);
        }

        String remaining = code.substring(index);
        String preview = remaining.length() > 20 ? remaining.substring(0, 20) : remaining;
        throw new IllegalStateException("Invalid token near '" + preview + "' at index " + index);
    }

    boolean hasMoreTokens() {
        return skipIgnored(index) < code.length();
    }

    Token advance() {
        Token token = getCurrentToken();
        index += token.getValue().length();
        return token;
    }

    Token eat(String token) {
        if (hasMoreTokens() && getCurrentToken().getValue().equals(token)) {
            return advance();
        }

        throw new IllegalStateException("Invalid token near '" + token + "' at index " + index);
    }

    TokenType getTokenType() {
        return getCurrentToken().getType();
    }

    String getStringValue() {
        Token token = getCurrentToken();
        if (token.getType() != TokenType.STRING_CONSTANT) {
            throw new IllegalStateException("You can't perform this action on non string value");
        }

        return token.getValue().substring(1, token.getValue().length() - 1);
    }

    int getIntValue() {
        Token token = getCurrentToken();
        if (token.getType() != TokenType.INTEGER_CONSTANT) {
            throw new IllegalStateException("You can't perform this action on non int value");
        }

        return Integer.parseInt(token.getValue());
    }

    String getKeyWord() {
        Token token = getCurrentToken();
        if (token.getType() != TokenType.KEYWORD) {
            throw new IllegalStateException("You can't perform this action on non keyword value");
        }

        return token.getValue();
    }

    char getSymbol() {
        Token token = getCurrentToken();
        if (token.getType() != TokenType.SYMBOL) {
            throw new IllegalStateException("You can't perform this action on symbol value");
        }

        return token.getValue().charAt(0);
    }

    String getIdentifier() {
        Token token = getCurrentToken();
        if (token.getType() != TokenType.IDENTIFIER) {
            throw new IllegalStateException("You can't perform this action on non identifier");
        }

        return token.getValue();
    }

    private int skipIgnored(int currentIndex) {
        while (currentIndex < code.length()) {
            String current = code.substring(currentIndex);
            boolean matchedIgnoredToken = false;

            for (Pattern ignoredRule : IGNORED_RULES) {
                Matcher matcher = ignoredRule.matcher(current);
                if (!matcher.find()) {
                    continue;
                }

                currentIndex += matcher.group().length();
                matchedIgnoredToken = true;
                break;
            }

            if (!matchedIgnoredToken) {
                break;
            }
        }

        return currentIndex;
    }

    private void validateToken(TokenType type, String token) {
        if (type != TokenType.INTEGER_CONSTANT) {
            return;
        }

        try {
            int number = Integer.parseInt(token);
            if (number > 32767) {
                throw new IllegalStateException(
                    "Integer constant out of range near '" + token + "' at index " + index);
            }
        } catch (NumberFormatException exception) {
            throw new IllegalStateException(
                "Integer constant out of range near '" + token + "' at index " + index);
        }
    }

    private static final class Rule {
        private final Pattern pattern;
        private final TokenType type;

        private Rule(Pattern pattern, TokenType type) {
            this.pattern = pattern;
            this.type = type;
        }
    }
}
